extensions/replication/tasks/MultipleBackendTask.js

```diff
@@ -370,21 +333,73 @@ class MultipleBackendTask extends ReplicateObject {
 
+    /**
+     * Get a byte range size for an object of the given content length, such
+     * that the range count does not exceed 1024 elements (i.e.
+     * MPU_GCP_MAX_PARTS).
+     * @param {Number} contentLen - The content length of the whole object
+     * @return {Number} The range size to use
+     */
+    _getGCPRangeSize(contentLen) {
+        const pow = Math.pow(2, Math.ceil(Math.log(contentLen) / Math.log(2)));
+        const rangeSize = Math.ceil(pow / MPU_GCP_MAX_PARTS);
+        return rangeSize;
+    }
+
+    /**
+     * Get a byte range size for an object of the given content length, such
+     * that the range size sums to the content length when multiplied by a value
+     * between 2 and 1000, excepting a very large object. This has the effect of
+     * creating MPU parts of the given range size. This method also optimizes
+     * for the subsequent range requests by returning a range size that is an
+     * interval of MB or GB.
+     * @param {Number} contentLen - The content length of the whole object
+     * @return {Number} The range size to use
+     */
+    _getRangeSize(contentLen) {
+        let rangeSize = (1024 * 1024) * 16; // Default 16MB part size.
+        if (contentLen <= rangeSize) {
+            return contentLen;
+        }
+        // Target creation of an MPU that is between a 2 and 1000 parts.
+        while (contentLen / rangeSize > 1000) {
+            // When given a very large object we want to allow use of up to 10K
+            // parts, so we limit the part size to 512MB here.
+            if (rangeSize >= (1024 * 1024) * 512) {
+                break;
+            }
+            rangeSize *= 2;
+        }
+        // If the object is large enough to exceed 10K parts of 512MB, then at
+        // this point we need to increase the part size such that the largest
+        // object size of 5TB can be accounted for.
+        while (contentLen / rangeSize > 10000) {
+            rangeSize *= 2;
+        }
+        return rangeSize;
+    }
+
     /**
      * Get byte ranges for an object of the given content length, such that the
-     * range count does not exceed 1024 elements.
+     * range count does not exceed 1024 parts if replicating to GCP or does not
+     * exceed 10000 parts otherwise.
      * @param {Number} contentLen - The content length of the whole object
+     * @param {Boolean} isGCP - Whether the object is being replicated to GCP
      * @return {Array} The array of byte ranges.
      */
-    _getRanges(contentLen) {
-        // Creates ranges that have sizes that are powers of 2.
-        const pow = Math.pow(2, Math.ceil(Math.log(contentLen) / Math.log(2)));
-        const range = Math.ceil(pow / MPU_GCP_MAX_PARTS);
+    _getRanges(contentLen, isGCP) {
+        if (contentLen === 0) {
+            // 0-byte object has no range. However we still want to put a single
+            // part so the range in the subsequent GET request is undefined.
+            return [undefined];
```

Comment:
One small change. Can we use `null` here instead?

Reply:
Updated.